tests/ConnectionTest.php

```diff
@@ -120,6 +120,58 @@ public function adusted_max_attempts_value_is_applied()
         }
     }
 
+    /**
+     * @test
+     */
+    public function does_not_reconnect_if_transaction_active()
+    {
+        $connection = $this->mockedConnection(null, 'server has gone away');
+        $connection->setMaxAttempts(5);
+        $connection->beginTransaction();
+        
+        try {
+            $connection->runQuery('select * from users');
+        } catch (ConnectionException $exception) {
+            $this->assertEquals(1, $exception->getAttempts());
+        }
+    }
+
+    /**
+     * @test
+     *
+     * If an error occurs during a transaction, an exception will be thrown.
+     * However, once the exception has been handled, it's not impossible that
+     * more queries may be executed. To that end, the failure of a transaction
+     * should not interfere with any future queries.
+     */
+    public function transaction_failure_does_not_affect_future_queries()
+    {
+        $mock = new PDOExceptionThrower('sqlite::memory:');
+
+        $connection = new Connection(fn () => $mock);
+
+        $connection->setMaxAttempts(5);
+
+        // First we'll initiate a transaction - which should fail.
+
+        try {
+            $mock->throwOnQuery('(not a connection issue)');
+            $connection->beginTransaction();
+            $connection->runQuery('select * from users');
+        } catch (PDOException $exception) {
+            // (We just want to suppress this.)
```

Comment:
worth asserting `expectException`? (just a thought, no biggy)

Reply:
Ultimately it's a good point; it would be nice to have some assurance that `runQuery` failed. I've added an assertion to reflect this, but not using `expectException`.

(The problem with the `expectException` approach is that the rest of the assertions become unreachable i.e. the exception will be caught by PHPUnit and then the test ends.)

Reply:
> (The problem with the expectException approach is that the rest of the assertions become unreachable

great point!

Reply:
> Is it too late to remove our dependency on this package?

Rather than just removing this, I think we'd benefit from at least replacing it with something else that does the same job. As questionable as this implementation may be, it does ultimately serve a good purpose (Laravel's connection manager does the same thing under the hood, for example).

Reply:
I didn't realise time was of the essence on this. But I see your point about the need for it, it's just not the sort of thing we should be maintaining ourselves.

Reply:
I agree. Unfortunately, I don't think there are any better (realistic) options out there. All of the other PDO wrappers I could find suffer from the same issue of not accounting for failed transactions - and we don't have the option of just dropping in something like Laravel's query class (because of the legacy code's hard dependency on PDO).

_An_ option might be creating a PDO wrapper which proxies requests _to_ something like Laravel's DB code? Would still be hacky and bloated, but would offload some of the logic. TBD in the future, maybe.